### ISSUE TITLE
OledDisplay class needs configuration added to init

### ISSUE THREAD

**Issue by stephenhensley** [electro-smith/libDaisy#166]
Currently it always uses SPI1, and two specific pins for DC and RST. It's also specifically setup for SPI Mode on this SSD1309 display.

having some sort of init struct or something that we pass in like

```C++

class OledDisplay
{

// . . .

struct OledDisplayInit
{
    OledCommOption hcomm; // where this could be OLED_COMM_I2C1, OLED_COMM_SPI1, etc.
    dsy_gpio_pin reset_pin, dc_pin; // standard gpio {port, pin} struct for the gpio.
};

// . . . 

void Init(OledDisplayInit *init);

// . . .

};
```

Open to other options, too.

**Comment by TheSlowGrowth** [comment]
Suggestion: Make it a template and provide the transport layer as a template argument.

    template <typename TransportLayer>
    class OledDisplay
    {
    public:
        Init(TransportLayer& tl) : tl_(tl) {}
        // ....
        void swapBuffersAndTransmit()
        {
            tl_.writeCommand(....);
            tl_.writeData(....);
         }
    private:
        TransportLayer& tl_;
    };

Then we can provide a default transport layer for I2C and SPI for users to use:

    SpiTranportLayer transportLayer;
    OledDisplay<SpiTransportLayer> display;
    transportLayer.Init(...);
    display.Init(transportLayer);

To make user code even simpler, we could even go as far as providing specialisations for I2C and SPI that already include their tranport layers like this 

    class OledDisplaySpi public: OledDisplay<SpiTransportLayer>
    {
    public:
        Init(spi_config spi)
        {
            spi_transport_layer_.Init(spi);
            OledDisplay::Init(spi_transport_layer_);
        }
    private:
        SpiTransportLayer spi_transport_layer_;
    }

This leaves options open for the users. They can use the simple to use specialisations or provide their own transport layer (e.g. for parallel interfaces, etc.).
I know inheritance seems like overhead in an embedded system, but here, we don't even use virtual functions. I expect no overhead compared to a "all in one class" implementation. As a benefit, we get better expandability.

**Comment by TheSlowGrowth** [comment]
I'll write such a driver for a SH1106 display. If you like it and decide to include it in libDaisy, I can also adjust the existing SSD1309 driver for such a model.

**Comment by recursinging** [comment]
What's the status on this issue? I've a I2C SSD1306 in my seed based kxmx_bluemchen eurorack module, and the suggestion from @TheSlowGrowth would be very helpful for me to stay better aligned with libDaisy and the Patch module.

**Comment by stephenhensley** [comment]
Well, snap. I missed the comments on this until the most recent one.

I've been slacking on getting this together, but I do like the transport layer idea. I wasn't totally sure on how we would want to handle i2c vs spi for this type of thing.

If @TheSlowGrowth has made the sh1106 module already, I'd be happy to add it, and assist in the modification of the existing ssd1309 (which afaik also works for the ssd1306).

**Comment by StaffanMelin** [comment]
I think it would be great to add I2C support for the OledDisplay class. The I2C and ssd1306 combination is very common and seems to be the cheapest and simplest to get your hands om. I did some work on that combo but never managed to get it to work, unfortunately.

**Comment by stephenhensley** [comment]
resolved in #326 

Thanks again @recursinging !
